### PR TITLE
[Group 1] Enable nullable annotations for `Microsoft.Extensions.Primitives`

### LINKED ISSUE #43605: Annotate Microsoft.Extensions.* assemblies for nullable reference types
In .NET 6.0, we plan to annotate Microsoft.Extensions.* assemblies built from the dotnet/runtime repo. Work involves:

* Submit individual PRs, one for each assembly. Each PR should include changes to both the src and the ref. Each PR should contain only changes related to the nullable annotations/attributes, no other changes.
* PRs can be merged once the annotations have been appropriately reviewed in PR.

*In .NET 5.0, nullable references types was tracked in #2339, and remainder are being tracked in #41720 for .NET 6.0*

#### Group 1 ####
- [ ] Microsoft.Extensions.Logging.Abstractions
- [ ] Microsoft.Extensions.DependencyModel
- [ ] Microsoft.Extensions.DependencyInjection.Abstractions
- [ ] Microsoft.Extensions.Primitives
- [ ] Microsoft.Extensions.FileSystemGlobbing
- [ ] Microsoft.Extensions.Configuration.Abstractions
- [ ] Microsoft.Extensions.FileProviders.Abstractions

#### Group 2 ####
- [ ] Microsoft.Extensions.Options
- [ ] Microsoft.Extensions.Caching.Abstractions
- [ ] Microsoft.Extensions.DependencyInjection
- [ ] Microsoft.Extensions.FileProviders.Physical
- [ ] Microsoft.Extensions.FileProviders.Composite
- [ ] Microsoft.Extensions.Hosting.Abstractions
- [ ] Microsoft.Extensions.Configuration
- [ ] Microsoft.Extensions.Configuration.FileExtensions 
- [ ] Microsoft.Extensions.Configuration.Ini
- [ ] Microsoft.Extensions.Configuration.Json
- [ ] Microsoft.Extensions.Configuration.Xml *(Check first if `System.Security.Cryptography.Xml` nullable enabled)*
- [ ] Microsoft.Extensions.Configuration.CommandLine
- [ ] Microsoft.Extensions.Configuration.Binder *(Check first if `System.ComponentModel.TypeConverter` nullable enabled)*
- [ ] Microsoft.Extensions.Configuration.EnvironmentVariables
- [ ] Microsoft.Extensions.Configuration.UserSecrets

#### Group 3  ####
- [ ] Microsoft.Extensions.Logging
- [ ] Microsoft.Extensions.Http
- [ ] Microsoft.Extensions.Caching.Memory
- [ ] Microsoft.Extensions.Options.DataAnnotation
- [ ] Microsoft.Extensions.Options.ConfigurationExtensions

#### Group 4  ####
- [ ] Microsoft.Extensions.Logging.Configuration
- [ ] Microsoft.Extensions.Logging.EventLog *(Check first if `System.Diagnostics.EventLog` nullable enabled)*
- [ ] Microsoft.Extensions.Logging.EventSource
- [ ] Microsoft.Extensions.Logging.TraceSource *(Check first if `System.Diagnostics.TraceSource` nullable enabled)*
- [ ] Microsoft.Extensions.Logging.Debug

#### Group 5 ####
- [ ] Microsoft.Extensions.Logging.Console
- [ ] Microsoft.Extensions.Hosting

### PR DESCRIPTION
**Pull request by maxkoshevoi**
Related to #43605

### CODE REVIEW

File: src/libraries/Microsoft.Extensions.Primitives/ref/Microsoft.Extensions.Primitives.cs [hunk 1]
```diff
@@ -116,49 +117,49 @@ public partial struct Enumerator : System.Collections.Generic.IEnumerator<Micros
             public void Reset() { }
         }
     }
-    public readonly partial struct StringValues : System.Collections.Generic.ICollection<string>, System.Collections.Generic.IEnumerable<string>, System.Collections.Generic.IList<string>, System.Collections.Generic.IReadOnlyCollection<string>, System.Collections.Generic.IReadOnlyList<string>, System.Collections.IEnumerable, System.IEquatable<Microsoft.Extensions.Primitives.StringValues>, System.IEquatable<string>, System.IEquatable<string[]>
+    public readonly partial struct StringValues : System.Collections.Generic.ICollection<string>, System.Collections.Generic.IEnumerable<string>, System.Collections.Generic.IList<string>, System.Collections.Generic.IReadOnlyCollection<string>, System.Collections.Generic.IReadOnlyList<string>, System.Collections.IEnumerable, System.IEquatable<Microsoft.Extensions.Primitives.StringValues>, System.IEquatable<string?>, System.IEquatable<string[]>
```
Comment (stephentoub) [inline src/libraries/Microsoft.Extensions.Primitives/ref/Microsoft.Extensions.Primitives.cs:120]: Should more of these be `string?`?  Are you not allowed to add nulls for example?

File: src/libraries/Microsoft.Extensions.Primitives/ref/Microsoft.Extensions.Primitives.cs [hunk 2]
```diff
@@ -58,8 +59,8 @@ public partial interface IChangeToken
         public bool Equals(Microsoft.Extensions.Primitives.StringSegment other) { throw null; }
         public static bool Equals(Microsoft.Extensions.Primitives.StringSegment a, Microsoft.Extensions.Primitives.StringSegment b, System.StringComparison comparisonType) { throw null; }
         public bool Equals(Microsoft.Extensions.Primitives.StringSegment other, System.StringComparison comparisonType) { throw null; }
-        public override bool Equals(object obj) { throw null; }
-        public bool Equals(string text) { throw null; }
+        public override bool Equals(object? obj) { throw null; }
+        public bool Equals(string? text) { throw null; }
```
Comment (stephentoub) [inline src/libraries/Microsoft.Extensions.Primitives/ref/Microsoft.Extensions.Primitives.cs:63]: Doesn't this throw for null?

File: src/libraries/Microsoft.Extensions.Primitives/ref/Microsoft.Extensions.Primitives.cs [hunk 3]
```diff
@@ -116,49 +117,49 @@ public partial struct Enumerator : System.Collections.Generic.IEnumerator<Micros
...
+    public readonly partial struct StringValues : System.Collections.Generic.ICollection<string>, System.Collections.Generic.IEnumerable<string>, System.Collections.Generic.IList<string>, System.Collections.Generic.IReadOnlyCollection<string>, System.Collections.Generic.IReadOnlyList<string>, System.Collections.IEnumerable, System.IEquatable<Microsoft.Extensions.Primitives.StringValues>, System.IEquatable<string?>, System.IEquatable<string[]>
     {
         private readonly object _dummy;
         private readonly int _dummyPrimitive;
         public static readonly Microsoft.Extensions.Primitives.StringValues Empty;
-        public StringValues(string value) { throw null; }
+        public StringValues(string? value) { throw null; }
         public StringValues(string[] values) { throw null; }
```
[the remark this answers is not in the document]
Comment (maxkoshevoi) [reply src/libraries/Microsoft.Extensions.Primitives/ref/Microsoft.Extensions.Primitives.cs:126]: Yes, from what I understood. `StringValues` can contain: `null`, `string`, or `string[]`.

[the remark this answers is not in the document]
Comment (stephentoub) [reply src/libraries/Microsoft.Extensions.Primitives/ref/Microsoft.Extensions.Primitives.cs:126]: This works fine:
```C#
var sv = new StringValues(new string?[] { "a", "b", null, "c" });
Console.WriteLine(sv.ToString());
```

File: src/libraries/Microsoft.Extensions.Primitives/ref/Microsoft.Extensions.Primitives.cs (hunk 2 above)
Comment (maxkoshevoi) [reply src/libraries/Microsoft.Extensions.Primitives/ref/Microsoft.Extensions.Primitives.cs:63]: It doesn't allow me to make it non-nullable...

image

Comment (maxkoshevoi) [reply src/libraries/Microsoft.Extensions.Primitives/ref/Microsoft.Extensions.Primitives.cs:63]: Interestingly, in implementation `text` is non-nullable and everything compiles fine

Comment (stephentoub) [reply src/libraries/Microsoft.Extensions.Primitives/ref/Microsoft.Extensions.Primitives.cs:63]: > It doesn't allow me to make it non-nullable...

You need to change the type of the interface being implemented to `IEquatable<string?>`.

> in implementation text is non-nullable and everything compiles fine

Yes, because the code never tries to dereference a maybe-null value.  The first thing the code does is check whether it's null and throw if it is.

File: src/libraries/Microsoft.Extensions.Primitives/ref/Microsoft.Extensions.Primitives.cs (hunk 3 above)
[the remark this answers is not in the document]
Comment (maxkoshevoi) [reply src/libraries/Microsoft.Extensions.Primitives/ref/Microsoft.Extensions.Primitives.cs:126]: Actually, I don't see a reason, why `StringValues` cannot contain `string?[]`. I'll update the annotations accordingly.

File: src/libraries/Microsoft.Extensions.Primitives/ref/Microsoft.Extensions.Primitives.cs [hunk 4]
```diff
@@ -58,8 +59,8 @@ public partial interface IChangeToken
         public bool Equals(Microsoft.Extensions.Primitives.StringSegment other) { throw null; }
         public static bool Equals(Microsoft.Extensions.Primitives.StringSegment a, Microsoft.Extensions.Primitives.StringSegment b, System.StringComparison comparisonType) { throw null; }
         public bool Equals(Microsoft.Extensions.Primitives.StringSegment other, System.StringComparison comparisonType) { throw null; }
-        public override bool Equals(object obj) { throw null; }
-        public bool Equals(string text) { throw null; }
+        public override bool Equals(object? obj) { throw null; }
```
[the remark this answers is not in the document]
Comment (maxkoshevoi) [reply src/libraries/Microsoft.Extensions.Primitives/ref/Microsoft.Extensions.Primitives.cs:62]: Thanks for the tip! This might mon actually be the case in this case. Since `Equals` will return `true` if `StringValue` contains `null` or empty string.

[the remark this answers is not in the document]
Comment (stephentoub) [reply src/libraries/Microsoft.Extensions.Primitives/ref/Microsoft.Extensions.Primitives.cs:62]: This is StringSegment, not StringValue.  StringSegment.Equals(object) is defined as:
```C#
       public override bool Equals(object obj)
        {
            return obj is StringSegment segment && Equals(segment);
        }
```

[the remark this answers is not in the document]
Comment (maxkoshevoi) [reply src/libraries/Microsoft.Extensions.Primitives/ref/Microsoft.Extensions.Primitives.cs:62]: I think the only overload where this is applicable is this one. I'll add it there.

image

[the remark this answers is not in the document]
Comment (maxkoshevoi) [reply src/libraries/Microsoft.Extensions.Primitives/ref/Microsoft.Extensions.Primitives.cs:62]: Sorry, didn't notice the type you were commenting on

File: src/libraries/Microsoft.Extensions.Primitives/ref/Microsoft.Extensions.Primitives.cs (hunk 2 above)
Comment (maxkoshevoi) [reply src/libraries/Microsoft.Extensions.Primitives/ref/Microsoft.Extensions.Primitives.cs:63]: > You need to change the type of the interface being implemented to IEquatable<string?>

Still doesn't allow non-nullable text

File: src/libraries/Microsoft.Extensions.Primitives/ref/Microsoft.Extensions.Primitives.cs [hunk 5]
```diff
@@ -58,8 +59,8 @@ public partial interface IChangeToken
...
+        public bool Equals(string? text) { throw null; }
         public bool Equals(string text, System.StringComparison comparisonType) { throw null; }
```
[the remark this answers is not in the document]
Comment (maxkoshevoi) [reply src/libraries/Microsoft.Extensions.Primitives/ref/Microsoft.Extensions.Primitives.cs:64]: No, it throws on when text is null (same as for `Equals(string? text)`, but it has to be nullable there as discussed in different comment).

File: src/libraries/Microsoft.Extensions.Primitives/ref/Microsoft.Extensions.Primitives.cs (hunk 2 above)
Comment (stephentoub) [reply src/libraries/Microsoft.Extensions.Primitives/ref/Microsoft.Extensions.Primitives.cs:63]: Sorry, I misread what you'd written.  `IEquatable<T>` is defined to expect `Equals` always allows null (it accepts a `T?`), which is why this is complaining about trying to specify a non-nullable value.  I think the implementation should be changed to return false rather than throwing, and then typed as `string?`, but I'll defer here to @dotnet/area-microsoft-extensions area owners.

File: src/libraries/Microsoft.Extensions.Primitives/ref/Microsoft.Extensions.Primitives.cs (hunk 4 above)
[the remark this answers is not in the document]
Comment (maxkoshevoi) [reply src/libraries/Microsoft.Extensions.Primitives/ref/Microsoft.Extensions.Primitives.cs]: Fixed

File: src/libraries/Microsoft.Extensions.Primitives/ref/Microsoft.Extensions.Primitives.cs (hunk 3 above)
[the remark this answers is not in the document]
Comment (maxkoshevoi) [reply src/libraries/Microsoft.Extensions.Primitives/ref/Microsoft.Extensions.Primitives.cs]: Fixed

File: src/libraries/Microsoft.Extensions.Primitives/ref/Microsoft.Extensions.Primitives.cs [hunk 6]
```diff
@@ -116,49 +117,49 @@ public partial struct Enumerator : System.Collections.Generic.IEnumerator<Micros
...
+        public static bool operator ==(string? left, Microsoft.Extensions.Primitives.StringValues right) { throw null; }
         public static bool operator ==(string[] left, Microsoft.Extensions.Primitives.StringValues right) { throw null; }
-        public static implicit operator string (Microsoft.Extensions.Primitives.StringValues values) { throw null; }
-        public static implicit operator string[] (Microsoft.Extensions.Primitives.StringValues value) { throw null; }
-        public static implicit operator Microsoft.Extensions.Primitives.StringValues (string value) { throw null; }
-        public static implicit operator Microsoft.Extensions.Primitives.StringValues (string[] values) { throw null; }
+        public static implicit operator string?(Microsoft.Extensions.Primitives.StringValues values) { throw null; }
+        public static implicit operator string[](Microsoft.Extensions.Primitives.StringValues value) { throw null; }
+        public static implicit operator Microsoft.Extensions.Primitives.StringValues(string? value) { throw null; }
+        public static implicit operator Microsoft.Extensions.Primitives.StringValues(string[] values) { throw null; }
```
[the remark this answers is not in the document]
Comment (maxkoshevoi) [reply src/libraries/Microsoft.Extensions.Primitives/ref/Microsoft.Extensions.Primitives.cs]: Fixed

File: src/libraries/Microsoft.Extensions.Primitives/ref/Microsoft.Extensions.Primitives.cs (hunk 1 above)
Comment (maxkoshevoi) [reply src/libraries/Microsoft.Extensions.Primitives/ref/Microsoft.Extensions.Primitives.cs]: Fixed

File: src/libraries/Microsoft.Extensions.Primitives/ref/Microsoft.Extensions.Primitives.cs (hunk 2 above)
Comment (maxkoshevoi) [reply src/libraries/Microsoft.Extensions.Primitives/ref/Microsoft.Extensions.Primitives.cs]: Made it return false.

File: src/libraries/Microsoft.Extensions.Primitives/ref/Microsoft.Extensions.Primitives.cs (hunk 5 above)
[the remark this answers is not in the document]
Comment (maxkoshevoi) [reply src/libraries/Microsoft.Extensions.Primitives/ref/Microsoft.Extensions.Primitives.cs]: Made it nullable & return false instead of throwing.

File: src/libraries/Microsoft.Extensions.Primitives/src/StringTokenizer.cs [hunk 7]
```diff
@@ -1,7 +1,6 @@
 // Licensed to the .NET Foundation under one or more agreements.
 // The .NET Foundation licenses this file to you under the MIT license.
 
-using System;
```
Comment (eerhardt) [inline src/libraries/Microsoft.Extensions.Primitives/src/StringTokenizer.cs:4]: Can you undo this unnecessary change? That way it doesn't pollute source control history.

File: src/libraries/Microsoft.Extensions.Primitives/src/CompositeChangeToken.cs [hunk 8]
```diff
@@ -102,10 +108,10 @@ private void EnsureCallbacksInitialized()
             }
         }
 
-        private static void OnChange(object state)
+        private static void OnChange(object? state)
         {
-            var compositeChangeTokenState = (CompositeChangeToken)state;
-            if (compositeChangeTokenState._cancellationTokenSource == null)
+            var compositeChangeTokenState = (CompositeChangeToken?)state;
+            if (compositeChangeTokenState?._cancellationTokenSource == null)
```
Comment (eerhardt) [inline src/libraries/Microsoft.Extensions.Primitives/src/CompositeChangeToken.cs:114]: This is actually changing behavior since there is a new `null` check getting inserted here. I think it would be better to `Debug.Assert` that `state != null` since the only callers of this method pass `this` for `state`.

File: src/libraries/Microsoft.Extensions.Primitives/src/ChangeToken.cs [hunk 9]
```diff
@@ -100,7 +100,7 @@ private void RegisterChangeTokenCallback(IChangeToken token)
                     return;
                 }
 
-                IDisposable registraton = token.RegisterChangeCallback(s => ((ChangeTokenRegistration<TState>)s).OnChangeTokenFired(), this);
+                IDisposable registraton = token.RegisterChangeCallback(s => ((ChangeTokenRegistration<TState>?)s)?.OnChangeTokenFired(), this);
```
Comment (eerhardt) [inline src/libraries/Microsoft.Extensions.Primitives/src/ChangeToken.cs:103]: Similar comment here as in `CompositeChangeToken.cs` - this is adding a null check where it isn't necessary. `this` will be passed to this method, so `s` will never be `null`. We can add a `!` here to silence the warning.

File: src/libraries/Microsoft.Extensions.Primitives/src/ChangeToken.cs [hunk 10]
```diff
@@ -142,7 +142,7 @@ public void Dispose()
             {
                 // If the previous value is disposable then dispose it, otherwise,
                 // now we've set the disposed sentinel
-                Interlocked.Exchange(ref _disposable, _disposedSentinel).Dispose();
+                Interlocked.Exchange(ref _disposable, _disposedSentinel)?.Dispose();
```
Comment (eerhardt) [inline src/libraries/Microsoft.Extensions.Primitives/src/ChangeToken.cs:145]: I don't believe `_disposable` can ever be `null`, and thus this should never need to check for `null`.

The ctor of this class calls `RegisterChangeTokenCallback`, which calls `SetDisposable`, which always ensures `_disposable` is non-null.

Comment (eerhardt) [reply src/libraries/Microsoft.Extensions.Primitives/src/ChangeToken.cs:145]: My mistake - looking more at this, I see `RegisterChangeTokenCallback` checks `if (token is null)`, but that doesn't match the nullabillity of the `IChangeToken token` parameter. Does the signature of `RegisterChangeTokenCallback` need to be updated to allow for a `null` token?

Comment (maxkoshevoi) [reply src/libraries/Microsoft.Extensions.Primitives/src/ChangeToken.cs:145]: Does it make sense to call `ChangeToken.OnChange` and not subscribe to a change? Considering `RegisterChangeTokenCallback` doesn't throw an exception, maybe it is.

File: src/libraries/Microsoft.Extensions.Primitives/src/StringSegment.cs [hunk 11]
```diff
@@ -254,11 +256,11 @@ public bool Equals(string text)
         /// <paramref name="text"/> is <see langword="null" />.
```
Comment (eerhardt) [inline src/libraries/Microsoft.Extensions.Primitives/src/StringSegment.cs:256]: This exception doc should be updated.

File: src/libraries/Microsoft.Extensions.Primitives/src/ChangeToken.cs [hunk 12]
```diff
@@ -142,7 +142,7 @@ public void Dispose()
             {
                 // If the previous value is disposable then dispose it, otherwise,
                 // now we've set the disposed sentinel
-                Interlocked.Exchange(ref _disposable, _disposedSentinel).Dispose();
+                Interlocked.Exchange(ref _disposable, _disposedSentinel)!.Dispose();
```
Comment (eerhardt) [inline src/libraries/Microsoft.Extensions.Primitives/src/ChangeToken.cs:145]: Sorry - I was confusing on my previous comment. I think what you had before was correct:

```suggestion
                Interlocked.Exchange(ref _disposable, _disposedSentinel)?.Dispose();
```

It is possible for `_disposable` to be `null` here - when someone creates this class with a `changeTokenProducer` that returns `null`. In that case, nothing will call `SetDisposable` (as far as I can tell), and `_disposable` will still be null when this class is Disposed.

Can you add a unit test for this scenario while we are here? Basically the test is:

```C#
ChangeToken.OnChange(() => null, () => Debug.Fail("Should never be called")).Dispose();
```

File: src/libraries/Microsoft.Extensions.Primitives/src/ChangeToken.cs (hunk 10 above)
Comment (krwq) [reply src/libraries/Microsoft.Extensions.Primitives/src/ChangeToken.cs:145]: I'd prefer we did product changes separately mainly for two reasons:
- less chance to add any regression - in here you'll not regress since you're just adding a null check but it opens a window of temptation for more and more complex tiny refactoring which could regress and it's easier to just not allow any product changes in nullability PR and focus on checking annotations rather than "is the product change correct" on top of nullable annotations
- null check PR will be a separate commit and will be easier to find if we search for "what changes have been made to Microsoft.Extensions.Primitives" - this is better in case we need to backport some behavior or tracking some

File: src/libraries/Microsoft.Extensions.Primitives/src/CompositeChangeToken.cs [hunk 13]
```diff
@@ -13,11 +14,14 @@ namespace Microsoft.Extensions.Primitives
...
-        private CancellationTokenSource _cancellationTokenSource;
-        private bool _registeredCallbackProxy;
-        private List<IDisposable> _disposables;
+        private static readonly Action<object?> _onChangeDelegate = OnChange;
+        private readonly object _callbackLock = new();
+        private CancellationTokenSource? _cancellationTokenSource;
+        private List<IDisposable>? _disposables;
+
+        [MemberNotNullWhen(true, nameof(_cancellationTokenSource))]
+        [MemberNotNullWhen(true, nameof(_disposables))]
```
Comment (krwq) [inline src/libraries/Microsoft.Extensions.Primitives/src/CompositeChangeToken.cs:23]: I don't believe this will make any difference for the compiler (at least it doesn't on a test app I just tried) - also you're not doing anything to guarantee this in here. cc: @stephentoub (in case I'm wrong or there is something new coming)

File: src/libraries/Microsoft.Extensions.Primitives/src/CompositeChangeToken.cs [hunk 14]
```diff
@@ -13,11 +14,14 @@ namespace Microsoft.Extensions.Primitives
...
+        private bool _registeredCallbackProxy { get; set; }
```
Comment (krwq) [inline src/libraries/Microsoft.Extensions.Primitives/src/CompositeChangeToken.cs:24]: also please re-order back the fields

File: src/libraries/Microsoft.Extensions.Primitives/src/CompositeChangeToken.cs (hunk 13 above)
Comment (krwq) [reply src/libraries/Microsoft.Extensions.Primitives/src/CompositeChangeToken.cs:23]: actually I tried a bit different pattern than you have here and that one didn't work but specific case you have here works...

Comment (maxkoshevoi) [reply src/libraries/Microsoft.Extensions.Primitives/src/CompositeChangeToken.cs:23]: It's needed for `EnsureCallbacksInitialized` method to work. Since we returning from it if `_registeredCallbackProxy` is `true`, compiler complains that nothing is initialized.

And because `_registeredCallbackProxy` is set to `true` only an the end of that method, it can be used as a sign that those fields are initialized.

File: src/libraries/Microsoft.Extensions.Primitives/src/CompositeChangeToken.cs (hunk 14 above)
Comment (maxkoshevoi) [reply src/libraries/Microsoft.Extensions.Primitives/src/CompositeChangeToken.cs:24]: But this is property now.. Do i need to move it back with the fields?

File: src/libraries/Microsoft.Extensions.Primitives/src/CompositeChangeToken.cs (hunk 13 above)
Comment (krwq) [reply src/libraries/Microsoft.Extensions.Primitives/src/CompositeChangeToken.cs:23]: https://github.com/dotnet/roslyn/issues/55664

Comment (krwq) [reply src/libraries/Microsoft.Extensions.Primitives/src/CompositeChangeToken.cs:23]: yes, noticed that, the behavior I described seems to be a bug I referenced above

File: src/libraries/Microsoft.Extensions.Primitives/src/CompositeChangeToken.cs (hunk 14 above)
Comment (eerhardt) [reply src/libraries/Microsoft.Extensions.Primitives/src/CompositeChangeToken.cs:24]: Since it is a property, it shouldn't be named like a field.

File: src/libraries/Microsoft.Extensions.Primitives/src/StringValues.cs [hunk 15]
```diff
@@ -39,7 +40,7 @@ public StringValues(string value)
         /// Initializes a new instance of the <see cref="StringValues"/> structure using the specified array of strings.
         /// </summary>
         /// <param name="values">A string array.</param>
-        public StringValues(string[] values)
+        public StringValues(string?[] values)
```
Comment (krwq) [inline src/libraries/Microsoft.Extensions.Primitives/src/StringValues.cs:43]: should this be `string?[]? values`?

File: src/libraries/Microsoft.Extensions.Primitives/src/StringValues.cs [hunk 16]
```diff
@@ -57,7 +58,7 @@ public StringValues(string[] values)
         /// Defines an implicit conversion of a given string array to a <see cref="StringValues"/>.
         /// </summary>
         /// <param name="values">A string array to implicitly convert.</param>
-        public static implicit operator StringValues(string[] values)
+        public static implicit operator StringValues(string?[] values)
```
Comment (krwq) [inline src/libraries/Microsoft.Extensions.Primitives/src/StringValues.cs:61]: `string?[]?` ?

File: src/libraries/Microsoft.Extensions.Primitives/src/StringValues.cs [hunk 17]
```diff
@@ -581,59 +579,59 @@ public static bool Equals(StringValues left, StringValues right)
...
+        public bool Equals(string? other) => Equals(this, new StringValues(other));
 
         /// <summary>
         /// Determines whether the specified string array and <see cref="StringValues"/> objects have the same values.
         /// </summary>
         /// <param name="left">The string array to compare.</param>
         /// <param name="right">The <see cref="StringValues"/> to compare.</param>
         /// <returns><c>true</c> if the value of <paramref name="left"/> is the same as the value of <paramref name="right"/>; otherwise, <c>false</c>.</returns>
-        public static bool Equals(string[] left, StringValues right) => Equals(new StringValues(left), right);
+        public static bool Equals(string?[] left, StringValues right) => Equals(new StringValues(left), right);
```
Comment (krwq) [inline src/libraries/Microsoft.Extensions.Primitives/src/StringValues.cs:605]: should arrays be nullable?

Comment (krwq) [reply src/libraries/Microsoft.Extensions.Primitives/src/StringValues.cs:605]: here and anywhere in this file where applies

File: src/libraries/Microsoft.Extensions.Primitives/tests/ChangeTokenTest.cs [hunk 18]
```diff
@@ -234,6 +235,12 @@ public void DoubleDisposeDisposesOnce()
             Assert.Equal(2, provider.DisposeCalls);
         }
 
+        [Fact]
+        public void NullTokenDisposeShouldNotThrow()
+        {
+            ChangeToken.OnChange(() => null, () => Debug.Fail("Should never be called")).Dispose();
```
Comment (krwq) [inline src/libraries/Microsoft.Extensions.Primitives/tests/ChangeTokenTest.cs:241]: I'd recommend to use Assert.True(false) instead - Debug.Fail will stop the entire process rather than failing a single test case

File: src/libraries/Microsoft.Extensions.Primitives/tests/StringSegmentTest.cs [hunk 19]
```diff
@@ -480,17 +482,6 @@ public void StringSegment_Equals_String_Valid(string candidate, StringComparison
             Assert.Equal(expectedResult, result);
         }
 
-        [Fact]
```
[the remark this answers is not in the document]
Comment (maxkoshevoi) [reply src/libraries/Microsoft.Extensions.Primitives/tests/StringSegmentTest.cs:483]: It's no longer valid (see added lines above). `Equals(null)` now returns `false` instead of throwing.
Discussed here: https://github.com/dotnet/runtime/pull/57395#discussion_r688980555

[the remark this answers is not in the document]
Comment (krwq) [reply src/libraries/Microsoft.Extensions.Primitives/tests/StringSegmentTest.cs:483]: Sounds fine with me

**Review by krwq — APPROVED** [review]
LGTM, thank you a lot for this contribution!

**Review by eerhardt — APPROVED** [review]
LGTM. Thanks for the contribution!

**Comment by BrennanConroy** [comment]
I think StringSegment needs another look, Buffer is allowed to be null so you can distinguish between string.Empty and a null string.
You can see that null is a valid value because there is code specific to Buffer being null
https://github.com/dotnet/runtime/blob/8f2408cbfe16d0fc5da901e5caf68d0d59f33f39/src/libraries/Microsoft.Extensions.Primitives/src/StringSegment.cs#L86
https://github.com/dotnet/runtime/blob/8f2408cbfe16d0fc5da901e5caf68d0d59f33f39/src/libraries/Microsoft.Extensions.Primitives/src/StringSegment.cs#L30
https://github.com/dotnet/runtime/blob/8f2408cbfe16d0fc5da901e5caf68d0d59f33f39/src/libraries/Microsoft.Extensions.Primitives/src/StringSegment.cs#L80

This is causing issues in AspNetCore when trying to consume the Runtime update because we expect to be able to pass null to StringSegment. https://github.com/dotnet/aspnetcore/pull/35547

**Comment by maxkoshevoi** [comment]
Oh, I see. Second constructor thrown me off (it throws if `buffer` is `null`, but that's because we are trying to get only part of it).
I'll create a PR to make `buffer` nullable shortly.

**Comment by maxkoshevoi** [comment]
Shouldn't `new StringSegment(null).Equals((string?)null);` return `true` than?
In only implementation it throws, and after this PR it returns `false` (which might need to be changed back to throwing now that `buffer` can be null).
